apps/emqx_utils/src/emqx_utils_stream.erl

```diff
@@ -93,6 +95,33 @@ map(F, S) ->
         end
     end.
 
+%% @doc Make a stream by filtering the underlying stream with a predicate function.
+filter(F, S) ->
+    FilterNext = fun FilterNext(St) ->
+        case emqx_utils_stream:next(St) of
+            [X | Rest] ->
+                case F(X) of
+                    true ->
+                        [X | filter(F, Rest)];
+                    false ->
+                        FilterNext(Rest)
+                end;
+            [] ->
+                []
+        end
+    end,
+    fun() -> FilterNext(S) end.
```

Comment:
Q: Worth to simplify to?
```suggestion
    fun FilterNext(F, S) ->
        case next(S) of
            [X | Rest] ->
                case F(X) of
                    true ->
                        [X | filter(F, Rest)];
                    false ->
                        FilterNext(F, Rest)
                end;
            [] ->
                []
        end
    end.
```

Reply:
🤔 this gives a 2-arity function while stream is a 0-arity one

Reply:
Oh right. Disregard please.